tests/unit/test_schema_merger.py

```diff
@@ -133,9 +133,20 @@ def should_successfully_merge_array_types_with(
                 ),
                 id="fields-of-different-type",
             ),
+            pytest.param(
+                StructType([StructField("a_field", UserDefinedType())]),
+                StructType([StructField("a_field", UserDefinedType())]),
+                pytest.raises(
+                    ValueError,
+                    match=re.escape(
+                        "Data type is not mergeable, expected one of: ['StructType', 'ArrayType', 'AtomicType'] but got 'UserDefinedType'"
```

Comment:
I haven't seen much use of it in the wild out there, the only reason I used it here because it was outside the options and I didn't feel there was going to be support for it soon, the how is maybe that users would define how they would merge it to sparkql (e.g. an abstract class for defining type X)

Side side note: there are other options that might be worth adding, like NullType, which is not AtomicType but should be treated similarly?
<img width="774" alt="Screenshot 2023-08-02 at 00 27 07" src="https://github.com/mattjw/sparkql/assets/7058741/8241e74b-83f2-440e-a177-47ceb2075cdc">